### PR TITLE
Replace linear_constraint and linear_expression with an MIT licensed implementation

### PR DESCRIPTION
**Pull request by dthaler**
This change addresses the next file of those listed in issue #149.

Signed-off-by: Dave Thaler <dthaler@ntdev.microsoft.com>

### CODE REVIEW

**Comment by dthaler** [comment]
Windows failure is again due to "Could NOT find Boost (missing: Boost_INCLUDE_DIR)" unrelated to these changes.  Will file a separate issue to track.

**Comment by dthaler** [comment]
Rebased to verify that all checks now pass.

File: src/crab/linear_constraint.hpp [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        _expression = expression;
+        _constraint_kind = constraint_kind;
+    }
+
+    [[nodiscard]] const linear_expression_t& expression() const { return _expression; }
+    [[nodiscard]] constraint_kind_t kind() const { return _constraint_kind; }
+
+    // Test whether the constraint is guaranteed to be false.
+    [[nodiscard]] bool is_contradiction() const {
+        if (!_expression.variable_terms().empty()) {
```
Comment (elazarg) [inline src/crab/linear_constraint.hpp:34]: Personally I prefer `!_expression.is_constant()`.

File: src/crab/linear_constraint.hpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    } else {
+        // Display constraint as the simpler form of (e.g.):
+        //     Ax + By < -C
+        // instead of the internal representation of:
+        //     Ax + By + C < 0
+        auto expression = constraint.expression();
+        expression.output_variable_terms(o);
+
+        const char* constraint_kind_label[] = {" == ", " <= ", " < ", " != "};
+        o << constraint_kind_label[(int)constraint.kind()] << -expression.constant_term();
```
Comment (elazarg) [inline src/crab/linear_constraint.hpp:87]: Isn't this risky? One day someone adds new constraint kind, and suddenly you get buffer overflow.

File: src/crab/linear_expression.hpp [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+    // Allow a caller to access individual terms.
+    const variable_terms_t& variable_terms() const { return _variable_terms; }
+    const number_t& constant_term() const { return _constant_term; }
+
+    // Multiply a linear expression by a constant.
+    [[nodiscard]] linear_expression_t operator*(const number_t& constant) const {
+        variable_terms_t variable_terms;
+        for (const auto& [variable, coefficient] : _variable_terms) {
+            variable_terms[variable] = coefficient * constant;
```
Comment (elazarg) [inline src/crab/linear_expression.hpp:52]: I think `.insert()` is better suited here.

**Review by elazarg — APPROVED** [review]
I like it. I left several comments, but all are minor and most reflect my own taste (or ignorance). I can merge as it is, just let me know when it's ready.

I'd really like @caballa to review this, especially the redundant check removed from split_dbm.cpp, but maybe a post-merge review would do.

File: src/crab/linear_constraint.hpp (hunk 2 above)
Comment (dthaler) [reply src/crab/linear_constraint.hpp]: Added check

File: src/crab/linear_expression.hpp (hunk 3 above)
Comment (dthaler) [reply src/crab/linear_expression.hpp]: After reading https://stackoverflow.com/questions/17172080/insert-vs-emplace-vs-operator-in-c-map, I updated it to use emplace()

File: src/crab/linear_expression.hpp [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        if (new_coefficient != 0) {
+            variable_terms[variable] = new_coefficient;
+        }
+        return linear_expression_t(variable_terms, _constant_term);
+    }
+
+    // Subtract one expression from another.
+    [[nodiscard]] linear_expression_t operator-(const linear_expression_t& expression) const {
+        variable_terms_t variable_terms = _variable_terms;
+        for (const auto& term : expression.variable_terms()) {
```
[the remark this answers is not in the document]
Comment (dthaler) [reply src/crab/linear_expression.hpp]: Done

File: src/crab/linear_constraint.hpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    NOT_ZERO
+};
+
+class linear_constraint_t final {
+  private:
+    linear_expression_t _expression = linear_expression_t(0);
+    constraint_kind_t _constraint_kind;
+
+  public:
+    linear_constraint_t(const linear_expression_t expression, constraint_kind_t constraint_kind)
```
[the remark this answers is not in the document]
Comment (dthaler) [reply src/crab/linear_constraint.hpp]: Updated

File: src/crab/split_dbm.hpp [hunk 6]
```diff
@@ -133,24 +133,24 @@ class SplitDBM final {
...
     Wt eval_expression(const linear_expression_t& e, bool overflow) {
-        Wt res(convert_NtoW(e.constant(), overflow));
+        Wt res(convert_NtoW(e.constant_term(), overflow));
         if (overflow) {
             return Wt(0);
         }
 
-        for (auto [n, v] : e) {
-            Wt coef = convert_NtoW(v, overflow);
+        for (auto [variable, coefficient] : e.variable_terms()) {
```
[the remark this answers is not in the document]
Comment (dthaler) [reply src/crab/split_dbm.hpp]: Updated anyway

File: src/crab/linear_expression.hpp [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        if (new_coefficient != 0) {
+            variable_terms[variable] = new_coefficient;
+        }
+        return linear_expression_t(variable_terms, _constant_term);
+    }
+
+    // Add two expressions.
+    [[nodiscard]] linear_expression_t operator+(const linear_expression_t& expression) const {
+        variable_terms_t variable_terms = _variable_terms;
+        for (const auto& term : expression.variable_terms()) {
```
[the remark this answers is not in the document]
Comment (dthaler) [reply src/crab/linear_expression.hpp]: Updated

File: src/crab/linear_constraint.hpp (hunk 1 above)
Comment (dthaler) [reply src/crab/linear_constraint.hpp]: Ok, added

**Comment by dthaler** [comment]
For future reference, https://en.wikipedia.org/wiki/Linear_equation was the "spec" used when implementing this code.  It was pretty straightforward. :)

**Pull request merged**